debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java

```diff
@@ -93,24 +176,45 @@ public R apply(final R record) {
             logger.warn("Expected Envelope for transformation, passing it unchanged");
             return record;
         }
-        final R newRecord = delegate.apply(record);
-        if (newRecord.value() == null && dropDeletes) {
-            logger.trace("Delete message {} requested to be dropped", record.key());
-            return null;
+        final R newRecord = afterDelegate.apply(record);
+        if (newRecord.value() == null) {
+            // Handling delete records
+            switch (handleDeletes) {
+                case DROP:
+                    logger.trace("Delete message {} requested to be dropped", record.key());
+                    return null;
+                case REWRITE:
+                    logger.trace("Delete message {} requested to be rewritten", record.key());
+                    final R oldRecord = beforeDelegate.apply(record);
+                    return removedDelegate.apply(oldRecord);
+                default:
+                    return newRecord;
+            }
+        } else {
+            // Handling insert and update records
+            switch (handleDeletes) {
+                case REWRITE:
+                    logger.trace("Delete message {} requested to be rewritten", record.key());
```

Comment:
This log message is a little bit misleading.

Reply:
Right. Too much copy&paste